pkg/reconciler/managed/metrics.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package managed
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	kmetrics "k8s.io/component-base/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+func init() { //nolint:gochecknoinits // metrics should be registered once
+	metrics.Registry.MustRegister(drift, mr, mrReady, mrSynced, mrDetected, mrReadyDuration, mrDeletionDuration)
+}
+
+const subSystem = "crossplane"
+
+var (
+	drift = prometheus.NewHistogramVec(prometheus.HistogramOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
+		Subsystem: subSystem,
+		Name:      "resource_drift_seconds",
+		Help:      "ALPHA: How long since the previous successful reconcile when a resource was found to be out of sync; excludes restart of the provider",
+		Buckets:   kmetrics.ExponentialBuckets(10e-9, 10, 10),
+	}, []string{"group", "kind"})
+
+	mr = prometheus.NewGaugeVec(prometheus.GaugeOpts{ //nolint:gochecknoglobals // metrics should be registered once in init
```

Comment:
Is it possible to use counters, rather than gauges, to track the number of MRs?

Here's a good explanation of why: https://youtu.be/eq4CnIzw-pE?si=LAhhDrgrqBtrcacy&t=1008

In short, counters are nice because they don't lose meaning when sampled. Gauges can.

If we want to answer the question "how many MRs exist?" we can do that most accurately with two counters:

```
crossplane_managed_resources_created_total = 42
crossplane_managed_resources_deleted_total = 10
```

42 - 10 = 32 MRs exist.

You can also answer questions like "how many MRs were created over the last month" accurately, which you can't do with a gauge - we'd miss 10 MRs being created then deleted between sampling points.

The downside of counters is they don't work so well for questions like "how many MRs are currently ready?" when MRs can transition back and forth between being ready and unready (or being deleted).

We might want to consider a combination of counters and gauges. For example:

```
crossplane_managed_resources_created_total = 42
crossplane_managed_resources_deleted_total = 10
crossplane_managed_resources{condition_status_ready="True",condition_status_synced="True"} = 30
```

Reply:
Maybe I'm missing something, but again it's the same problem I've been mentioning about. It's hard to decide when to increase the counter metric in a reconciliation loop.